Step3. 볼링 점수판(점수 계산)

nara1030 opened the pull request:
정완님 안녕하세요.🙇 피드백 주신 걸 빨리 적용해보고 싶었는데 쉽지가 않네요...😅 시간이 너무 딜레이되는 것 같아 한 번 더 조언을 받고자 PR을 드립니다.

1. 피드백 관련하여 말씀해주신 상태 패턴 관련하여 적용을 해보았습니다. 혼자 생각해내지 못하고 [다음 글](https://dev-monkey-dugi.tistory.com/92)을 참고하여 기존 `Controller`와 `ResultView`의 if문을 제거해주었습니다. 다만 제대로 적용을 했는지는 조금 의문이 드는데요. 그 이유는 상태 객체가 갖는 필드가 달라 통일성이 없는 것 아닌가 생각이 들었습니다(`Frame` 객체 생성자에서 각 상태를 초기화하기 위한 인자가 다른 걸 보실 수 있습니다).

2. 기능 구현 관련하여는 계속 감을 못 잡았던 것이 10 프레임의 경우 이전 프레임과 상태가 다르다는 점과 상태에 따라 점수 계산 로직이 다르다는 것이었습니다(예를 들면 스트라이크 및 스페어의 경우 미래 시점에서 점수가 계산되어야 하는 것 때문에 계산 로직이 상태 로직에 포함되어야 하는지, `Frame` 객체에 포함되어야 하는지 감이 안 잡히더라구요). 사실 2단계 힌트를 뒤늦게 보았는데요. 외부에서 `Frame` 객체를 생성하는 게 아닌 `Frame` 객체를 `LinkedList`처럼 해야 계산이 된다는 뭔가 추상적인 생각이 뒤늦게 들었는데 제가 직접 이를 구현해본 적이 없어 그런지 게시글을 봐도 감이 잘 잡히지 않더라구요. 이미 외부에서 `Frame` 객체를 생성한 상태에서 리팩토링을 하려 했으나 if문만 늘어나고 코드가 점점 지저분해져서.. 그냥 질문을 드리는 게 낫다고 생각했습니다.

주저리주저리 길어졌는데, `Frame` 객체를 어떻게 변경해야 여기에서 스코어 계산이 가능해질지가 가장 큰 고민입니다.

읽어주셔서 감사합니다.  조언해주신 것에 비해 적용한 것이 너무 미미해 죄송스럽네요.😓 작은 조언이라도 말씀해주시면 감사하겠습니다.🙇 좋은 주말 되세요.⛲🌆

nara1030 committed:
docs: 요구사항 반영

nara1030 committed:
refactor: 도메인 외부에서 도메인 상태 사용하는 분기문 제거
- Controller 및 ResultView에서 도메인 상태 사용하는 분기문 제거(상태 패턴 도입)

nara1030 committed:
test: Frame 객체 테스트

nara1030 committed:
style: 변수 명명 수정

nara1030 committed:
refactor: 클래스명 변경

ghojeong commented on src/main/java/bowling/domain/Frame.java:
```diff
@@ -1,49 +1,85 @@
+    private final ThrowingState miss;
 
-    Frame() {
+    private ThrowingState currentFrameState;
```
Frame 을 StateContext 로 하셨군요!

ghojeong commented on src/main/java/bowling/domain/Frame.java:
```diff
@@ -1,49 +1,85 @@
 
-    private int first = BEFORE_BOWLING;
-    private int second = BEFORE_BOWLING;
+    private final ThrowingState ready;
+    private final ThrowingState firstBowl;
+    private final ThrowingState gutter;
+    private final ThrowingState strike;
+    private final ThrowingState secondBowl;
+    private final ThrowingState spare;
+    private final ThrowingState miss;
```
멤버로 선언해놓고 전혀 사용하고 있지 않는듯 한데,
굳이 남겨두고 있는 이유가 무엇인가요?
삭제해도 괜찮지 않을까 해서 여쭈어봅니다.

nara1030 replied:
아.. 생각지도 못했네요. 블로그를 따라 적다가 놓친 거 같습니다..ㅠㅠ

ghojeong commented on src/main/java/bowling/view/ResultView.java:
```diff
@@ -26,36 +26,6 @@ public static void printScore(String name, List<Frame> records) {
-        int first = frame.getFirst();
-        int second = frame.getSecond();
-
-        StringBuilder stringBuilder = new StringBuilder();
-        if (first != Frame.BEFORE_BOWLING) {
-            stringBuilder.append(convertFirstScore(first));
-        }
-        if (second != Frame.BEFORE_BOWLING) {
-            stringBuilder.append(convertSecondScore(first, second));
-        }
```
View 와 Controller 에서 if 문이 싹 제거되어서 속 시원하네요. 👍

nara1030 replied:
사실 도메인 내부로 if문을 옮겨준 거에 불과한 거 아닌가..(상태 패턴 클래스 `bowl()` 메소드에 if문 존재) 했지만 여력이 없어 넘겼습니다.🥴

ghojeong commented on src/test/java/bowling/domain/GameTest.java:
```diff
@@ -7,20 +7,20 @@
-
-        assertThat(given_game.play(2, 5)).isEqualTo(5);
-        assertThat(given_game.play(2, 5)).isEqualTo(5);
-    }
+//    private final Game given_game = new Game();
+//
+//    @Test
+//    void 게임_실행_비정상_핀갯수_초과() {
+//        assertThatThrownBy(() -> given_game.play(1, 11))
+//                .isInstanceOf(InvalidNumberOfFallenPinsException.class);
+//    }
+//
+//    @Test
+//    void 게임_실행_정상() {
+//        assertThat(given_game.play(1, 2)).isEqualTo(2);
+//        assertThat(given_game.play(1, 7)).isEqualTo(7);
+//
+//        assertThat(given_game.play(2, 5)).isEqualTo(5);
+//        assertThat(given_game.play(2, 5)).isEqualTo(5);
+//    }
 }
```
전부 주석 처리한 테스트 파일을 굳이 왜 남겨 놓으셨는지 여쭈어봅니다.
Step4 에서 작업을 하실 건가요?

nara1030 replied:
아.. 아니요. 사실 DM을 드릴까 하다가 중간에 너무 안 풀려서 그냥 구현 중간에 힌트를 얻고자 PR을 드렸습니다. 그냥 구현하다 말고 PR 요청드린 거라...😅

ghojeong commented on src/main/java/bowling/domain/Frames.java:
```diff
@@ -18,11 +18,16 @@ public class Game {
     }
 
-    public int play(final int round, final int numberOfFallenPins) {
+    public boolean playFrame(final int round, final int numberOfFallenPins) {
+        Frame frame = play(round, numberOfFallenPins);
+        return frame.isEnd();
+    }
+
+    private Frame play(final int round, final int numberOfFallenPins) {
```
개인적으로 round 는 Controller 에서 주입 받는 값이 아니라,
도메인 객체 내부에서 비즈니스 로직으로 처리되어야 하는 상태라고 생각합니다.
쓰러트린 볼링은 사용자 입력으로 결정될 수 밖에 없지만,
round 는 게임이 진행되면서 자동으로 바뀌어야 하지 않을까요?

즉 쓰러트린 핀의 갯수는 사용자에게 의존적이지만,
round 는 비즈니스 로직에 의해 결정되어야 할 도메인 상태라고 생각합니다.

nara1030 replied:
점수 계산 같은 경우만 봐도 round를 내부에서 생성해야 할 거 같아 링크드리스트를 생각하긴 했는데, 계속 뇌정지가 오네요.. 조금 더 생각해보겠습니다.😂

ghojeong replied:
참고로 링크드 리스트 아이디어 매우 좋습니다. 👍 
뇌 정지가 오실만큼 머리가 아프시겠지만, 휴식 취하시면서 진행해주세요.

ghojeong commented on src/main/java/bowling/domain/Frame.java:
```diff
@@ -1,49 +1,85 @@
+        if (currentFrameState instanceof Ready) {
             first = numberOfFallenPins;
-            return first;
+            changeState(currentFrameState);
+            return this;
         }
-        second = numberOfFallenPins;
-        return second;
+        if (currentFrameState instanceof RunningState) {
```
currentFrameState 에 따른 비즈니스 로직의 분기를
instanceof 를 통한 타입검사로 해결하고 있군요....

이럴거면 currentFrameState 를 enum 혹은 원래대로의 int 값으로 하는게 나을 것 같습니다.
미션을 진행하면서 "instanceof" 는 봉인되어서 사용하지 못한다고 생각해주세요.

nara1030 replied:
아... 너무 큼직한 문제들이 많다보니 도메인 내부의 if문은 생각도 못했네요. 다른 문제긴 한데 `instanceof` 연산자는 비권장되나요? 제가 일하는 곳에선 심심치 않게 본 거 같아서... SI라 그런가 싶네요.😐

ghojeong replied:
음 사실 instanceof 를 쓰는 코드가 많기는 합니다.
저는 위와 같은 로직을 "타입을 상태처럼 썼다" 라고 표현합니다.

Type 과 상태가 어떻게 다른지에 대해서 [지인에게 피드백](https://velog.io/@nathankim0/%EB%B9%84%EB%8F%99%EA%B8%B0-%EC%9E%91%EC%97%85-%EA%B4%80%EB%A6%AC-%ED%9A%8C%EA%B3%A0-Type-%EA%B3%BC-State) 을 드렸던 적이 있는데,
이 둘을 혼용해서 썼을때 생기는 끔찍함을 아직 몸소 경험해 보시지 않아서 어렵군요.

그냥 말로만 설명하면 엄홍준 님처럼 납득하기 쉽지 않은게 당연해서 그렇습니다.
일단 상태 패턴 적용만해도 충분히 어려울테니,
이 부분은 여력이 되실 때 리팩토링해주세요.

하고 싶은 말은 "클래스 타입은 상태로 구현되어서 상태처럼 쓸 수 있지만, 사실 타입과 상태는 다른 개념이다!!" 입니다.

ghojeong replied:
구현 쪽으로 도움이 되게 조언을 드리자면,
모든 instanceof 로직분기는
**polymorphism** 으로 대체할 수 있다고 조언드리고 싶습니다.

자바라는 언어가 특화되어 있는 기능중 하나 이므로 polymorphism 활용해주세요.

ghojeong commented on src/main/java/bowling/domain/state/RunningState.java:
```diff
@@ -0,0 +1,7 @@
+package bowling.domain.state;
+
+import bowling.domain.Frame;
+
+public interface RunningState extends ThrowingState {
+    Frame frame();
+}
```
`EndedState extends ThrowingState`
`RunningState extends ThrowingState`
`Ready implements RunningState`

RunningState 와 EndeState 를 나눈 의도가 잘 이해가 가지 않아서 여쭈어봅니다.
RunningState 에만 왜 frame 이라는 메서드가 필요한건가요?

nara1030 replied:
[블로그](https://dev-monkey-dugi.tistory.com/92) 글을 보고 아이디어를 얻었는데요. 게임 진행 시 프레임의 진행 혹은 종료 여부를 판단하기 위해 `ThrowingState` 아래에 `RunningState`와 `EndedState`를 놓는 게 괜찮겠다고 생각이 들었습니다([이 그림](https://github.com/nara1030/java-bowling/tree/step3#%EC%83%81%ED%83%9C-%ED%8C%A8%ED%84%B4-%EB%8F%84%EC%9E%85)을 참고해주세요). 단지 위에서 피드백 주신 `instanceof` 연산자를 사용해서 프레임 종료 여부를 판단해야겠다 생각은 했습니다만...

`RunningState`에만 `frame()` 메소드를 적은 이유는 `ResultView`에서 if문을 없애기 위해 상태 패턴의 `symbol` 메소드를 사용하고 싶었는데요. 예를 들어 Strike(10, X)를 한 묶음으로 생각하고 싶었습니다. 그런데 Spare나 Miss 같은 `EndedState`의 경우 이전 상태를 알아야 하겠더라구요. `Frame` 객체만을 상태로 갖는다면 currentState만을 알 수 있어서 `bowl()`하는 시점에서 previousState 자체를 주입해주었습니다. 그렇다보니 아래에 말씀드린 상태 객체가 갖는 통일성이 없어져 버렸습니다.

ghojeong commented on src/main/java/bowling/domain/state/SecondBowl.java:
```diff
@@ -0,0 +1,21 @@
+        throw new EndedFrameException();
+    }
+
+    @Override
+    public String symbol() {
+        return previousState.symbol() + "│" + previousState.frame().second();
+    }
```
RunningState 의 frame() 메서드가 사용되고 있는 유일한 순간인것 같은데,
이럴바에는 각 State 에서 previousState 를 멤버로 가지는게 아니라,
Frames 를 멤버로 가지고, 상태가 변화하면서 본인이 가진 Frames 를 다음 State 에게 넘겨주는게 낫지 않을까 하는 제안을 해봅니다.

nara1030 replied:
처음엔 `Frame`을 필드로 갖게 하고자 하였으나... Spare나 Miss 등의 경우 이전 상태를 기억해 함께 그려주어야 해야 했고 방법을 찾지 못해 비대칭적인 설계가 되어버린 거 같습니다.. 말씀하신대로 `Frames`를 필드로 가지는 방법을 고민해보겠습니다.

ghojeong replied:
>  말씀하신대로 Frames를 필드로 가지는 방법을 고민해보겠습니다.

Frames 를 필드로 가지도록 하겠다고 하셨는데,
종국에는 Frame 을 필드로 가지도록 하신 이유를 알 수 있을까요?

ghojeong commented on src/main/java/bowling/domain/state/FirstBowl.java:
```diff
@@ -0,0 +1,36 @@
+
+import bowling.domain.Frame;
+
+public class FirstBowl implements RunningState {
+    private Frame frame;
```
> 그 이유는 상태 객체가 갖는 필드가 달라 통일성이 없는 것 아닌가 생각이 들었습니다

우려하신 점에 동의 합니다.
저는 모든 State 는 Frames 를 가지도록 필드를 통일했으면 합니다.

반대로 Frame 에서는 ThrowingState 를 가지지 않았으면 합니다.

ghojeong commented on src/main/java/bowling/domain/Frame.java:
```diff
@@ -1,49 +1,85 @@
 
-    Frame() {
+    private ThrowingState currentFrameState;
+    private int first = READY;
+    private int second = READY;
```
제 생각에 Frame 은 StateContext 를 수행하기에는,
이미 first 와 second 상태를 가지고 있기에,
first 와 second 와 관련한 로직을 처리하기에 이미 차고 넘치는 역할을 가진것 같습니다.

Frame 이 아니라 별도의 객체로 ThrowingState 의 Context 객체를 만드시는 건 어떠신가요?

nara1030 replied:
너무 많은 필드를 갖고 있단 건 보였는데 감을 못 잡았네요. 어떻게 쪼갤 수 있을지 고민해보겠습니다.😅

ghojeong replied:
Frame 에 first 와 second 필드를 남겨두길 바랬는데,
아예 모든 필드를 없애버리셨군요. 😭 
값객체는 불변으로 갱신되더라도, 필드를 가졌으면 합니다.

ghojeong commented on src/main/java/bowling/domain/Frame.java:
```diff
@@ -1,49 +1,85 @@
     }
 
-    public int getFirst() {
+    private void changeState(ThrowingState throwingState) {
+        this.currentFrameState = throwingState.bowl();
+    }
```
일단 changeState 를 하는 방식은 무척 칭찬드리고 싶습니다.
어려운 부분인데 각 State 의 변경되는 로직은 깔끔하게 잘 구현해주셨네요. 👍

nara1030 replied:
감사합니다..😭

ghojeong commented on src/main/java/bowling/domain/Frames.java:
```diff
@@ -4,7 +4,7 @@
 import java.util.List;
 
-public class Game {
+public class Frames {
     public static final int MAP_ROUND_TO_INDEX_CONSTANT = 1;
 
     private static final List<Frame> CACHE;
```
> 10 프레임의 경우 이전 프레임과 상태가 다르다는 점과 상태에 따라 점수 계산 로직이 다르다는 것이었습니다.
> Frame 객체를 어떻게 변경해야 여기에서 스코어 계산이 가능해질지가 가장 큰 고민입니다.

제 생각에는 Frames 를 처음부터 10개로 초기화하지 말고, 한개한개씩 추가를 해나갔으면 합니다.
그럴 경우 마지막 Frame 이 가장 최신 프레임이고, previousFrame 을 마지막에서 그 이전 프레임을 통해 가져올 수 있을 것 같습니다.

각 State 가 Frames 를 갖도록 하고, 다음 State 로 넘어갈 때마다 동일한 Frames 를 생성자로 넘기도록 하면,
스코어 계산에 필요한 모든 정보를 State 가 가질 수 있도록 할 수 있습니다.

각 State 별로 달라지는 스코어 계산 비즈니스 로직은
State 객체에서 구현을 하면 될것 같고요.

nara1030 replied:
지금처럼 Frames을 for문을 통해 10개를 한 번에 만들지 말고 Frame이 종료되면 다음 Frame을 생성하여 추가하는 식으로 하란 말씀이시죠? 적어주신 부분 중 제가 헷갈리는 부분이,

> 그럴 경우 마지막 Frame 이 가장 최신 프레임이고, previousFrame 을 마지막에서 그 이전 프레임을 통해 가져올 수 있을 것 같습니다

인데요. 예를 들어,

```txt
이름   |   1   |   2   |   3   |   4
   A   |   x   |  4/5  |
       |  19   | 
   B   |   x   |   x   |   x
       |  30   |
```

위와 같이 1라운드의 점수 계산 시점이 다르고, 점수 계산은 1라운드부터 계산 가능 여부를 고려하여 계산해야 한다고 생각했는데... 이렇게 하면 if문을 사용해야 할 거 같아서요. 적어주신 부분은 계산을 앞이 아닌 뒤에서부터 해야 한다는 뜻인지요?

ghojeong replied:
넵 말씀주셨던 LinkedList 아이디어를 응용해본것인데,
마지막에 있는 Frame 을 가져오기 쉽도록 ArrayList 를 추천드립니다.

ghojeong replied:
참고로 이건, 저라면 이렇게 하지 않을까~ 하는 아이디어라서 꼭 따르시지 않아듀 좋습니다.

결국 구현하는 고생은 홍준님께서 하시게 되니까요.
이런 아이디어 부분은 꼭 따르지 않으시고, 본인이 편한 방식을 선택해주세요

ghojeong replied:
참고로 저는 Frame 이 first 와 second 를 가진 점수를 기록하는 값객체로 사용되고,
Frames 또한 그러한 값 객체들을 모아놓은 자료구조로 생각해서 드렸던 피드백이었습니다.

즉, 저는 Frame 과 Frames 에는  "계산 가능 여부"  판단 및 "계산" 을 하는 비즈니스 로직이 들어 있지 않고,
"계산 가능 여부" 판단 및 "계산" 은 전부 State 패턴의 State 로 구현될 수 있다고 생각했습니다.
if 문의 로직 분기는 전부 Polymorphism 으로 분기 가능하고요.

ghojeong replied:
> 제 생각에는 Frames 를 처음부터 10개로 초기화하지 말고, 한개한개씩 추가를 해나갔으면 합니다.

이 피드백은 홍준님의 주관을 너무 침해하는 것 같아서, 반영하지 않아도 됩니다.
지금 처럼 10개 만드는 방법도 저는 좋습니다!
구현에 조금더 힌트가 될까 해서, 저의 방식을 이야기해본 것입니다.

ghojeong reviewed (changes requested):
엄홍준 님!

상태 패턴 적용하기가 어려우셨을 텐데, 매우 능숙하게 상태가 변화해가는 로직을 구현해주셨습니다!
제 기준에서 Frame 이 지나치게 많은 역할과 필드를 갖고 있는 것 같습니다.
Frame 에는 first 와 second 멤버만 남겨두고,
상태 패턴의 Context 역할은 전혀 별도의 객체로 분리해서 구현을 해주시면 좋을 것 같습니다.

nara1030 committed:
test: 클래스명 수정

nara1030 committed:
refactor: 불필요 필드 제거

nara1030 committed:
refactor: 프레임 객체를 구성하는 각 볼링 객체 도입
- Bowl 객체 도입함으로써 각 시도별 상태 저장 가능
  (Frame 객체 필드 감소 및 값 객체로써 역할 충실)
- 상태 구분 불필요(RunningState, EndedState 객체 제거)

nara1030 committed:
feat: 10프레임 분리
- 10/3/10 볼링 시 이슈 발생
- 테스트 작성 미비

nara1030 committed:
feat: 프레임당 누적 점수 계산
- 10 프레임 점수 계산 이슈
- Frames 객체 너무 많은 기능 부과
  (점수 계산 메소드 보면 객체 지향 생활 체조 등 무시하고 구현 집중)
- ResultView.printBlank 제네릭 메소드 도입
- ResultView.accumulateScore 메소드 자바 8 도입 가능 여부 체크

nara1030 committed:
fix: 10프레임 표시 및 점수계산 이슈 수정
- 10/3/10, 10/3/2 와 같은 케이스 이슈 해결

nara1030 committed:
test: 볼링 테스트 추가
- 작성하고 보니 Bowl 객체에 대한 테스트가 아닌 Frame 객체에 대한
  테스트라 분리 필요...

nara1030 commented:
안녕하세요 정완님
기한이 많이 지났네요.^^; 절차지향적으로 기능만 돌아가는 느낌인데... 부족한대로 PR 드립니다.

1. `Frame` 객체를 Bowl 객체를 이용해 한 번 더 쪼갰습니다.
	- Frame 객체 자체의 상태가 줄었습니다.
	- Bowl 객체 하나에 스코어와 상태를 보관함으로써 기존에 있던 `RunningState`, `EndedState` 객체를 삭제해주었습니다.
2. 10프레임을 표현하기 위해 `Frame` 객체는 인터페이스를 구현하도록 수정했습니다.
3. `Frame` 객체의 `play` 메소드 실행 시 `instanceof` 연산자 사용하던 부분을 제거하였습니다.

- - -
이하는 제가 고민했던 부분입니다.

1. 일전에 `Frame` 객체는 값객체처럼 사용하면 어떨까 언급해주셨는데, 이 말은 결국 지금 제가 볼링시 `Frame` 객체의 `play`를 실행하는 것을 상태 객체로 위임하는 것을 말씀하신 걸까요?
2. 10프레임을 표현하기 위해 상태 객체(상태 패턴)를 임의로 추가해주었습니다(last 패키지 내부 클래스들). 처음에 설계하고 들어가지 않아 이렇게 되었는데, 이전 프레임들과 통일성을 가져야 하는 걸까요? 아니면 네이밍만 조금 가독성이 좋게 해주면 딱히 상관없을까요?
4. 스코어 계산의 경우 상태 객체로 위임하라고 말씀하셨는데, `Frame`을 갖고 있다보니 `Frames` 객체에서 계산해주게 되었습니다. 클래스가 굉장히 비대해졌는데.. 개선할 수 있는 방법이 있을까요?

감사합니다. 좋은 오후 되세요..^^

ghojeong commented on src/main/java/bowling/domain/Frame.java:
```diff
@@ -1,49 +1,13 @@
-    }
+import bowling.domain.state.ThrowingState;
+
+public interface Frame {
+    Frame play(final int round, final int numberOfFallenPins);
+    int first();
+    int second();
+    ThrowingState firstState();
+    ThrowingState secondState();
+    boolean firstOfFrame();
+    boolean endFrame();
```
> 일전에 Frame 객체는 값객체처럼 사용하면 어떨까 언급해주셨는데, 이 말은 결국 지금 제가 볼링시 Frame 객체의 play를 실행하는 것을 상태 객체로 위임하는 것을 말씀하신 걸까요?

Frame 을 값 객체처럼 사용하면 어떨까? 라고 말씀하신 부분은 제 의도와 일치합니다.

하지만 Frame 객체가 play 메서드를 갖게 되는 것은 제 의도와 전혀 일치하지 않습니다.
저는 play 메서드가 State 패턴에서의 Context 객체가 가지기를 원합니다.
이전에는 Context 가 Frames 였기 때문에, 값을 모아두는 Frames 컬렉션과,
상태 Context 를 관리하는 Context 객체를 분리하면 좋겠다고 조언 드린 것이었고요.

> 10프레임을 표현하기 위해 Frame 객체는 인터페이스를 구현하도록 수정했습니다.

지금 Frame 클래스는 play 메서드만을 가지고 상태를 전혀 가지지 않은, 값객체 조차 아닌 interface 로 보입니다.

10 프레임이라하면, 마지막 프레임인 것 같은데,
과거에 있었던 EndedState 만으로는 불충분한지 여쭈어봅니다.
아니면 하다못해 Frames 일급 컬렉션에서 ArrayList 의 size 가 10인 것을 체크하는 것으로는 불충분한지 여쭈어봅니다.

nara1030 replied:
1.

> 10 프레임이라하면, 마지막 프레임인 것 같은데,
> 과거에 있었던 EndedState 만으로는 불충분한지 여쭈어봅니다.

이 부분은 상태 객체에서 굳이 이런 구분이 필요하지 않을 거 같다고 조언해주셔서 제거해주고 생각하였습니다.

> 아니면 하다못해 Frames 일급 컬렉션에서 ArrayList 의 size 가 10인 것을 체크하는 것으로는 불충분한지 여쭈어봅니다.

이 부분은 프레임 종료 시 다음 프레임을 생성/추가해주는 식으로 코드를 작성하는 식으로 변경하여 프레임 외부에서 인덱스를 체크하는 것이 조금 어색하다고 느껴져 제외하였습니다.
**
기존: https://github.com/nara1030/java-bowling/blob/step2/src/main/java/bowling/domain/Game.java
변경: https://github.com/nara1030/java-bowling/blob/step3/src/main/java/bowling/domain/Frames.java

2. 위의 이유로 10프레임과 그 외 프레임을 구분하기 위하여 인터페이스를 사용할지, 혹은 추상 클래스를 사용할지를 고민했습니다. 딱히 어느 것을 사용하든 문제가 없을 거 같아 인터페이스를 만든 후 이를 구현하도록 하였습니다. 위에서 `Frame`이 인터페이스, 즉 메서드만 가지므로 값 객체가 아니라고 하셨는데 사실 `NormalFrame` 및 `LastFrame`을 위한 인터페이스라 저는 값 객체나 다름 없다고 생각하고 구현하였습니다. 하단에서 인터페이스를 List로 가지는 게 무슨 의민지 모르겠다고 하셨는데, 저는 `NormalFrame` 및 `LastFrame`을 모두 표현하기 위해 `private final List<Frame> frames = new ArrayList<>();` 이렇게 작성하였습니다.

3. 지적해주신대로 Context 객체, 즉 현재 구조로 보면 `Bowl` 객체가 `play` 역할을 담당해야 할 거 같습니다.  다만 궁금한 점은 현재 `Bowl` 객체의 `bowl()` 메소드가 상태를 변경하는 역할을 담당하고 있고, `Frame` 객체는 이를 이용하는 구조인데 이런 식으로 Context 객체의 상태를 바꾸는 메소드를 이용하는 것도 금지되는 건지 궁금합니다.

ghojeong commented on src/main/java/bowling/domain/Frames.java:
```diff
@@ -0,0 +1,206 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Frames {
+    public static final int MAP_ROUND_TO_INDEX_CONSTANT = 1;
+
+    private final List<Frame> frames = new ArrayList<>();
```
인터페이스를 List 로 만든게 무슨의미를 가지는지 전혀 이해가 가지 않습니다.

> Frame 객체 자체의 상태가 줄었습니다.

Frame 이 상태를 갖지 않도록 바랬던게 아니라, Frame 이 값객체로,
그리고 Frames 는 그러한 값들의 컬렉션을 관리하는 역할을 하기를 바랬습니다.

nara1030 replied:
위 댓글에 이유를 적어놓았습니다.

ghojeong replied:
확인했습니다!
덕분에 좀 파악이 되는군요

ghojeong commented on src/main/java/bowling/domain/Frames.java:
```diff
@@ -0,0 +1,206 @@
+        return frame.endFrame();
+    }
+
+    private void addFrame(final Frame frame) {
+        if (frames.size() == 0) {
+            frames.add(frame);
+        }
+        if (lastFrame().endFrame()) {
+            frames.add(frame);
+        }
+    }
+
+    private Frame play(final int numberOfFallenPins) {
+        return lastFrame().play(lastIndex(), numberOfFallenPins);
+    }
```
```suggestion
    public void addFrame(final Frame frame) {
        if (frames.size() == 0) {
            frames.add(frame);
        }
        if (lastFrame().endFrame()) {
            frames.add(frame);
        }
    }

    private Frame play(final int numberOfFallenPins) {
        return lastFrame().play(lastIndex(), numberOfFallenPins);
    }
```

하지만 play 를 통해 다음 Frame 을 생성하는 로직은,
이전에 Bowl 한 State 에 따라서 달라지게 된다고 생각합니다

저는 play 부분을 외부의 다른 객체에게 위임하고,
Frames 는 컬렉션의 기본적인 역할에 맞게, addFrame 메서드를 외부에 public 하게 공개하는 것으로 충분하지 않나? 라고 생각이 듭니다.

ghojeong commented on src/main/java/bowling/domain/Frames.java:
```diff
@@ -0,0 +1,206 @@
+        return player;
+    }
+
+    public List<String> score() {
+        List<String> scores = new ArrayList<>();
+        for (int round = 0; round < frames.size(); round++) {
+            int score = Bowl.READY;
+            // strike
+            score = scoreIfStrike(round);
+            if (score != Bowl.READY) {
+                scores.add(round, String.valueOf(score));
+                continue;
+            }
+            // spare
+            score = scoreIfSpare(round);
+            if (score != Bowl.READY) {
+                scores.add(round, String.valueOf(score));
+                continue;
+            }
+            // completed
+            score = scoreIfRemain(round);
+            if (score != Bowl.READY) {
+                scores.add(round, String.valueOf(score));
+                continue;
+            }
+            // not completed
+//            break;
+            scores.add(round, "");
+        }
+        return scores;
+    }
```
> 스코어 계산의 경우 상태 객체로 위임하라고 말씀하셨는데, Frame을 갖고 있다보니 Frames 객체에서 계산해주게 되었습니다. 클래스가 굉장히 비대해졌는데.. 개선할 수 있는 방법이 있을까요?

-1 값을 가지는 Bowl.READY 상수를 통해 READY 인지 확인하는 로직이 마음에 들지 않습니다.
근본적인 해결책은 Frames 에서 score 계산을 하지 않는 것입니다.
왜냐하면 조건을 분기하기 위해서 현재 무슨 State 인지 판단을 해야하기 떄문입니다.

상태 객체로 스코어 계산을 위임하기 어렵다면, 하다못해 상태 Context 객체로라도 옮겨주세요.
같은 if 문을 사용하더라도 -1 상수를 통해 조건을 확인하는 것은 허용하지 않겠습니다.

ghojeong commented on src/main/java/bowling/domain/state/normal/Ready.java:
```diff
@@ -0,0 +1,28 @@
+import bowling.domain.Frame;
+import bowling.domain.state.ThrowingState;
+
+public class Ready implements ThrowingState {
+    private Frame frame; // = new Frame(); // causes recursion issue
+
+    public Ready(Frame frame) {
+        this.frame = frame;
+    }
```
저는 각 상태들이 Frame"s" 일급 컬렉션을 갖기를 바랬는데,
그냥 Frame 객체를 넣은 것은 혹시 제가 피드백을 오해의 소지가 있게 전달드렸는지 궁금해서 여쭈어봅니다.

nara1030 replied:
아닙니다. 이 부분은 제대로 이해하였고, 처음엔 그렇게 구현하려고 하였는데요. 정확히 왜 저렇게 구현하였는지 정확히 기억이 나지 않네요.. 이 부분에 대한 수정부터 다시 진행해봐야겠습니다..

ghojeong commented on src/main/java/bowling/domain/state/last/LastReady.java:
```diff
@@ -0,0 +1,30 @@
+import bowling.domain.state.normal.Gutter;
+import bowling.domain.state.normal.Strike;
+
+public class LastReady implements ThrowingState {
+    private Frame frame;
+
+    public LastReady(Frame frame) {
+        this.frame = frame;
+    }
```
> 10프레임을 표현하기 위해 상태 객체(상태 패턴)를 임의로 추가해주었습니다(last 패키지 내부 클래스들). 처음에 설계하고 들어가지 않아 이렇게 되었는데, 이전 프레임들과 통일성을 가져야 하는 걸까요? 아니면 네이밍만 조금 가독성이 좋게 해주면 딱히 상관없을까요?

저는 last 패키지 내부 클래스들이 딱히 통일성이 없다고 생각이 들지 않았는데,
왜 통일성이 없다고 생각하셨는지 궁금하군요.
interface 의 강제에 따라서 저는 충분히 통일되었다고 느껴져서요.

nara1030 replied:
마지막 프레임과 그 이전 프레임들의 상태를 구분하지 않고 표현할 수 있지 않을까 하는 고민을 하다가 결국 못해서.. 그래서 통일성이 없다고 느꼈던 거 같습니다.

ghojeong commented on src/main/java/bowling/domain/Bowl.java:
```diff
@@ -0,0 +1,49 @@
+import bowling.domain.state.ThrowingState;
+import bowling.exception.InvalidNumberOfFallenPinsException;
+
+public class Bowl {
+    public static final int READY = -1;
+    public static final int MAX_NUMBER_OF_PIN = 10;
+
+    private ThrowingState throwingState;
+    private int numberOfFallenPins;
```
아무래도 현재 Bowl 이 상태 Context 인듯 한데,
하고 있는 일이 많이 적은것 같습니다.
Frames 에 지나치게 많다고 느껴지는 기능들을 우선 Bowl 에게 옮기는 것부터 해보셨으면 합니다.

ghojeong reviewed (approved):
엄홍준님!

제가 피드백을 미숙하게 드려서, 오히려 더 혼란스럽게 만들고 있는 것 같아 죄송스럽네요.
제가 원하는 것은 2가지 입니다.

1. State 디자인 패턴의 적용과, 유의미한 상태 Context 객체의 역할을 찾아내기
2. if 문을 상태 객체의 polymorphism 으로 대체해보기

피드백 열심히 반영해주셔서 머지하겠습니다.
어려운게 있으면 언제든 DM 주세요.
마지막까지 화이팅입니다. 🔥

nara1030 commented:
시간적인 한계가 있어 머지시켜주셨다고 느꼈는데, 제가 이해하지 못하고 그냥 구현만 하는 것 같아 중간에 멈춰버렸네요. 연휴 전에 정완님 코드를 한 번 살펴봤는데 객체지향적으로 잘 쪼개져있다는 느낌 정도 오고 이해하기 쉽지 않더라구요. 부족함을 정말 많이 느껴서 잠시 멈추었습니다. 개인적으로는 볼링이 마음에 들어서 어쨌든 완료할 때까지 꾸준히 생각하게 될 거 같은데... 약간 부침이 느껴져서 DM도 못 드렸네요. 좋은 피드백 주셔서 감사드리고, 어쨌든 말씀 주신 걸로 고민해보고 이해하게 되어 구현을 하게 되면 좋겠습니다..^^;

ghojeong commented:
홍준님이 마음에 들어서 허들을 높여 피드백을 드렸는데,
괜히 자신감만 꺾었을까봐 걱정되네요.

제 피드백과는 별개로 기능 동작이 되면 Step4 머지를 해드릴 수 있습니다.
어려우시면 어려운대로 Step4 도전하셔서 할수 있는데까지 하시면 좋겠습니다.

nara1030 commented:
주말도 지나고 해서 DM도 못 드렸는데 확인해주시다니 죄송하고 고맙습니다. 제 구현과는 별개로 피드백은 정말 좋았고, 자신감이 꺾였다기보다 개발자 하려면 이렇게 해야 하는구나 느낌을 좀 받았달까요..ㅋㅋ 객관화가 된 거 같아 괜찮았습니다. 감사합니다, 고민해보고 조금 시간이 걸리더라도 구현이 되면 슬랙으로 말씀드리겠습니다. 좋은 하루 되세요!
